### PR TITLE
jewel: rgw : fix race in RGWCompleteMultipart

### PR DESCRIPTION
**Pull request by mattbenjamin**
http://tracker.ceph.com/issues/20900

### CODE REVIEW

**Comment by mattbenjamin** [comment]
(this failed in teuthology rgw suite, reproduced by hand, working it)

**Comment by mattbenjamin** [comment]
@orit this is not much more inelegant than the change on master, which also did this inline in CompleteMultipart;  we should refactor upstream, but just need Jewel to be correct

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -4180,9 +4181,41 @@ void RGWCompleteMultipart::execute()
...
+  librados::IoCtx ioctx;
+  rados::cls::lock::Lock l("RGWCompleteMultipart");
+  int max_lock_secs_mp = s->cct->_conf->rgw_mp_lock_max_time;
+  utime_t time(max_lock_secs_mp, 0);
+
+  op.assert_exists();
+  l.set_duration(time);
+  l.lock_exclusive(&op);
+
+  op_ret = store->get_obj_ioctx(meta_obj, &ioctx);
```
Comment (oritwas) [inline src/rgw/rgw_op.cc]: can this be moved before the locking?

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -4180,9 +4181,41 @@ void RGWCompleteMultipart::execute()
...
+    op_ret = -ERR_INTERNAL_ERROR;
+    s->err.message = "This multipart completion is already in progress";
+    return;
+  }
+
   op_ret = get_obj_attrs(store, s, meta_obj, attrs);
   if (op_ret < 0) {
-    ldout(s->cct, 0) << "ERROR: failed to get obj attrs, obj=" << meta_obj
+    ldout(s->cct, 0) << "ERROR: failed to get obj attrs, obj="
+		     << meta_obj
```
Comment (oritwas) [inline src/rgw/rgw_op.cc]: nit

File: src/rgw/rgw_op.cc (hunk 1 above)
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc]: iiuc, you want to move the lock structure setup after get_obj_ioctx?  that's fine, but I don't think there's a critical section here

Comment (oritwas) [inline src/rgw/rgw_op.cc]: I agree

**Comment by smithfarm** [comment]
@oritwas This passed an rgw suite at http://tracker.ceph.com/issues/20613#note-53 and @mattbenjamin re-pushed. Can you take another look?

@mattbenjamin Can you include a "Conflicts" section at the end of the commit message?

**Comment by smithfarm** [comment]
@oritwas Have your concerns been addressed now?

**Comment by smithfarm** [comment]
This passed an rgw suite at http://tracker.ceph.com/issues/20613#note-63

**Pull request merged**